spring-cloud-circuitbreaker-commons/src/main/java/org/springframework/cloud/circuitbreaker/commons/ReactorCircuitBreaker.java

```diff
@@ -0,0 +1,47 @@
+/*
+ * Copyright 2013-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.cloud.circuitbreaker.commons;
+
+import java.util.function.Function;
+
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+
+/**
+ * Spring Cloud reactor circuit breaker API.
+ *
+ * @author Ryan Baxter
+ */
+public interface ReactorCircuitBreaker {
```

Comment:
Is it possible to make it a bit more generic? There are more types than Mono/Flux (e.g. ParallelFlux, or reactor-netty's, or user provided), and it would be better to be more generic here

Reply:
I think this gets at the crux of the change.  We though it might be a good idea to have an interface that supports wrapping a `Publisher` hence we introduces `ReactiveCircuitBreaker`.  We still wanted to keep this interface around as it seemed like just supporting a general `Publisher` made the usage a bit awkward.  

For example if I called `reativeCircuitBreaker.run(myFlux)` I would pass in a `Flux` get back a `Publisher` when what I really want back is a `Flux`.  I was unsure of how to deal with that.

Reply:
The best would be to use:
```
myFlux.transform(reativeCircuitBreaker::run)
```

`transform` applies `Flux.from` automatically, and it does not break the reactive chain

Reply:
But doesnt that seem like a pain for every user to have to do?

Reply:
they will have to apply it one way or another, right?
Just, IMO, with `.transform(cb::run)` it is a bit more "natural" and less imperative than `cb.run(myFlux.filter(...).map(...).whatever(...))`.

I'm curious to hear the opinions of other Reactor team members /cc @smaldini @simonbasle

Reply:
I tend to like the pattern of applying the circuit breaker through `flux.transform(cb::run)` better. That said, that's more of a general preference, as I'm not entirely familiar if it makes total sense within the cb API.

For the generic aspect of the interface (or rather the avoidance of redundant `run` methods for each Reactor reactive type), this might be an extra argument in favor of having a base interface common to `Flux`/`Mono`/`ConnectableFlux`/`ParrallelFlux`: it would enable generic signatures like `<T extends ReactorType> T run(T original)`...
But we don't have such a common interface, so...

Reply:
> But we don't have such a common interface

`CorePublisher` maybe? Although that's 3.3.x

Reply:
might work... not sure what the plan is for `spring-cloud-circuitbreaker` in terms of which minimum Reactor release train to depend on.

Reply:
ping @bsideup

Reply:
@bsideup @simonbasle this is planned to be part of out Spring Cloud Hoxton release which will be based on Spring Boot 2.2.x.  So whatever version of Reactor that is included there.....which I believe is 3.3.x,  So maybe we can use `CorePublisher`?

Reply:
That's great, then yeah `CorePublisher` could be used.